internal/daemon/controller/handlers/targets/target_service.go

```diff
@@ -224,50 +235,112 @@ func (s Service) ListTargets(ctx context.Context, req *pbs.ListTargetsRequest) (
 		return &pbs.ListTargetsResponse{}, nil
 	}
 
-	tl, err := s.listFromRepo(ctx, userPerms)
+	pageSize := int(s.maxPageSize)
+	// Use the requested page size only if it is smaller than
+	// the configured max.
+	if req.GetPageSize() != 0 && uint(req.GetPageSize()) < s.maxPageSize {
+		pageSize = int(req.GetPageSize())
+	}
+	filter, err := handlers.NewFilter(ctx, req.GetFilter())
 	if err != nil {
 		return nil, err
 	}
-	if len(tl) == 0 {
-		return &pbs.ListTargetsResponse{}, nil
-	}
 
-	filter, err := handlers.NewFilter(ctx, req.GetFilter())
+	repo, err := s.repoFn(target.WithPermissions(userPerms))
 	if err != nil {
 		return nil, err
 	}
+	filterItemFn := func(ctx context.Context, item target.Target) (bool, error) {
+		pbItem, err := toProto(ctx, item, newOutputOpts(ctx, item, authResults, authzScopes)...)
+		if err != nil {
+			return false, err
+		}
```

Comment:
This operation unfortunately has to be performed twice for each item - once when filtering while listing in the DB and once when converting the results to the protobuf result after listing is complete. If we had some way to convert the API level filter to the domain this would avoid us having to convert to a protobuf type here (and probably allow us to skip this function altogether), but I think such a large refactor would be outside the scope of this change.

Reply:
I added a TODO here, do you think we need to add something like a JIRA task?